fsspec/implementations/prefix.py

```diff
@@ -0,0 +1,135 @@
+import os
+from pathlib import Path
+from typing import Any, Iterable, Sequence, Union
+
+import fsspec
+from fsspec import AbstractFileSystem
+from fsspec.core import split_protocol
+from fsspec.utils import stringify_path
+
+
+class PrefixFileSystem(AbstractFileSystem):
+    def __init__(
+        self,
+        prefix: str,
+        filesystem: fsspec.AbstractFileSystem,
+        *args,
+        **storage_options,
+    ) -> None:
+        super().__init__(*args, **storage_options)
+        self.prefix = prefix
+        self.filesystem = filesystem
+
+    def _add_fs_prefix(self, path: Union[str, Path]) -> Union[str, Sequence[str]]:
+        if isinstance(path, (str, Path)):
+            path = stringify_path(path)
+            protocol, path = split_protocol(path)
+            path = os.path.join(self.prefix, path)
+            return protocol + "://" + path if protocol is not None else path
+        elif isinstance(path, Iterable):
+            return [self._add_fs_prefix(x) for x in path]
+        assert False
+
+    def _remove_fs_prefix(self, path: Union[str, Path]) -> Union[str, Sequence[str]]:
+        if isinstance(path, (str, Path)):
+            path = stringify_path(path)
+            protocol, path = split_protocol(path)
+            path = os.path.relpath(path, start=self.prefix)
```

Comment:
Might be missing something, but I don't thin we can use `os.path` here because this could be ran on windows and because `relpath` is slow (esp on windows and esp when dealing with large number of files, e.g. in `ls()`). We should be able to just do something like `path = path[len(self.prefix]:]` instead.

Reply:
I changed from `relpath` to your suggestion!